src/trace-plugin-interface.js

```diff
@@ -0,0 +1,242 @@
+'use strict';
+var cls = require('./cls.js');
+var constants = require('./constants.js');
+
+/**
+ * This file describes an interface for third-party plugins to enable tracing
+ * for arbitrary modules.
+ */
+
+/**
+ * An object that is associated with a single child span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function ChildSpan(agent, context) {
+  this.agent = agent;
+  this.context = context;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+ChildSpan.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+ChildSpan.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * An object that is associated with a single root span. It exposes
+ * functions for adding labels to or closing the associated span.
+ */
+function Transaction(agent, context) {
+  this.agent = agent;
+  this.context = context;
+  this.closed = false;
+}
+
+/**
+ * Adds a label to the underlying span.
+ * @param {string} key The name of the label to add.
+ * @param {*} value The value of the label to add.
+ */
+Transaction.prototype.addLabel = function(key, value) {
+  this.context.addLabel(key, value);
+};
+
+/**
+ * Ends the underlying span. This function should only be called once.
+ */
+Transaction.prototype.endSpan = function() {
+  this.context.close();
+};
+
+/**
+ * Runs the given function in a child span, passing it an object that
+ * exposes an interface for adding labels and closing the span.
+ * @param {object} options An object that specifies options for how the child
+ * span is created and propogated.
+ * @param {string} options.name The name to apply to the child span.
+ * @param {?function(string, string)} options.setHeader A function describing
```

Comment:
@ofrobots The hope is that `setHeader` and `getHeader` are always set in plugins when the concept of a header exists, i.o.w cross-service requests. So this applies to anything HTTP-related, including the HTTP core module itself. I agree that the documentation doesn't clearly detail exactly when or why this is necessary, so I'll try to come up with a better description.

The only example so far is this:

https://github.com/kjin/cloud-trace-nodejs/blob/b1e1ec4df89e8324de240375a493a53baa927bfb/test/fixtures/plugin-express.js#L20